tool2schema/parameter_schema.py

```diff
@@ -0,0 +1,165 @@
+        """
+        Add the enum property to the given schema.
+        """
+        pass
+
+    def _add_description(self, schema: dict):
+        """
+        Add the description of this parameter, extracted from the function docstring, to the given schema.
+        """
+        if self.docstring is None:
+            return
+
+        docstring = " ".join(
+            [x.strip() for x in self.docstring.replace("\n", " ").split()]
+        )
+        params = re.findall(r":param (.*?): (.*?);", docstring)
+        for name, desc in params:
+            if name == self.parameter.name:
+                schema["description"] = desc
+                return
+
+    def _add_default(self, schema: dict):
+        """
+        Add the default value, when present, to the given schema.
+        """
+        if self.parameter.default == Parameter.empty:
+            return
+
+        schema["default"] = self.parameter.default
+
+    def to_json(self) -> dict:
+        """
+        Return the json schema for this parameter.
+        """
+        json = {}
+        self._add_description(json)
+        self._add_default(json)
+        self._add_items(json)
+        self._add_type(json)
+        self._add_enum(json)
+        return json
+
+
+class ValueTypeSchema(ParameterSchema):
+    """
+    Parameter schema for value types.
+    """
+
+    @staticmethod
+    def matches(parameter: Parameter) -> bool:
+        return (
+            parameter.annotation != Parameter.empty
+            and parameter.annotation.__name__ in TYPE_MAP
+        )
+
+    def _add_type(self, schema: dict):
+        schema["type"] = TYPE_MAP[self.parameter.annotation.__name__]
+
+
+class GenericParameterSchema(ParameterSchema):
+    """
+    Base class for generic parameter types supporting subscription.
+    """
+
+    def _get_sub_type(self):
+        if args := typing.get_args(self.parameter.annotation):
+            return TYPE_MAP.get(args[0].__name__, "object")
+
+        return None
+
+
+class ListParameterSchema(GenericParameterSchema):
+    """
+    Parameter schema for list (array) types.
+    """
+
+    @staticmethod
+    def matches(parameter: Parameter) -> bool:
+        return parameter.annotation != Parameter.empty and (
+            parameter.annotation is list
+            or typing.get_origin(parameter.annotation) is list
+        )
+
+    def _add_type(self, schema: dict):
+        schema["type"] = TYPE_MAP["list"]
+
+    def _add_items(self, schema: dict):
+        if sub_type := super()._get_sub_type():
+            schema["items"] = {"type": sub_type}
+
+
+class OptionalParameterSchema(GenericParameterSchema):
+    """
+    Parameter schema for typing.Optional types.
+    """
+
+    @staticmethod
+    def matches(parameter: Parameter) -> bool:
+        return (
+            parameter.annotation != parameter.empty
+            and typing.get_origin(parameter.annotation) is typing.Union
+            and type(None) in typing.get_args(parameter.annotation)
```

Comment:
Do you think checking length of args here is important, given `typing.Optional` accepts only one type.

Reply:
Not sure I understand the question. In this case typing.get_args returns a tuple with two elements, because it sees the optional as a union `(None, Y)` where Y is some other type

Reply:
Or are you referring to the case when the optional is not subscripted?

Reply:
Do we want to support it? It would be like a parameter without a type hint
```python
# a is whatever
def f(a: typing.Optional):
    pass
```

Reply:
I don't we need to support that.

Sorry, let me rephrase my original question (which wasn't very clear 😓)

Should be check if the number of args is 2? Because a user could conceivably do

```python
def f(a: typing.Union[str, list, None])
```

We then would assume that it was an optional type of `str`.

Reply:
Got it, will do later

Reply:
Done